Reduxify Consortia View

swashcap opened the pull request:
#### asana
https://app.asana.com/0/45796090191289/53621870692570

This moves the “consortium” view into a Redux-centered architecture. Notable changes:

* Creation of consortium actions and action creators.
* Creation of a `Consortia` service that acts as an adapter to PouchDB storage
* Creation of a `consortium-single-controller` component that facilitates bound action creators and “dumbing down” of the `consortium-single` component.
* Creation of consortium middleware that interacts with the services and modifies actions as needed (only for errors right now).

cdaringe commented on app/js/actions/consortium.js:
```diff
@@ -0,0 +1,178 @@
+export const CONSORTIUM_ERROR = 'CONSORTIUM_ERROR';
+
+export function fetchConsortium(id) {
+    return dispatch => {
```
where does this dispatch get bound again? bindAction creators?  I'm unclear/forgetting how the actual dispatch function makes it into this.  right now it looks like dispatch is undefined

cdaringe commented on app/js/actions/consortium.js:
```diff
@@ -0,0 +1,178 @@
+ * Consortium retrieval.
+ */
+
+export const CONSORTIUM_RECIEVE = 'CONSORTIUM_RECIEVE';
```
RECEIVE

swashcap replied:
Shoot, good catch.

swashcap replied:
Fixed in 4c662e38b8e68e85c67decf4a2d32abafb6ffe75.

cdaringe commented on app/js/actions/consortium.js:
```diff
@@ -0,0 +1,178 @@
+        id,
+        type: CONSORTIUM_ADD_RESULTS,
+    };
+}
```
file looks great.  not sure what it means to ADD_RESULTS/EDIT_RESULTS, but perhaps i will discover shortly

swashcap replied:
I stubbed actions and action creators for every conceivable mutation of a consortium. These will come in handy when we add results (or “aggregrates”.)

swashcap replied:
These action creators’ signatures were adjusted to match the required arguments in the consortium service. See 8efc0c08ee18cc42f0bf4b50c4fd455cdc324144.

cdaringe commented on app/js/components/consortium-single-controller.js:
```diff
@@ -0,0 +1,80 @@
+
+class ConsortiumSingleController extends Component {
+    componentWillMount() {
+        const {
```
these don't look to be used

swashcap replied:
See https://github.com/MRN-Code/coinstac/pull/8/files#diff-142df8d1ba7457c09427ead8d2355320R64

cdaringe replied:
ah.  i'm still getting used to _finding_ the variable in the nest.  thanks.

cdaringe commented on app/js/components/consortium-single-controller.js:
```diff
@@ -0,0 +1,80 @@
+    }
+    render() {
+        const { consortium, actions } = this.props;
+        const validate = partial(validateAnalysis, consortium._id);
```
i need to get on board with the partial train.  :rage4: rad

cdaringe commented on app/js/components/consortium-single-controller.js:
```diff
@@ -0,0 +1,80 @@
+ * Wire up action creators with the right parameters (usually consortium ID.)
+ * This keeps the sub-components simpler.
+ */
+function mapDispatchToProps(dispatch, ownProps) {
```
pre-loading all of the actionCreators like this seems like a really great idea.  wow, nice thinkin!

cdaringe commented on app/js/components/consortium-single.js:
```diff
@@ -1,200 +1,191 @@
+        this.toggleShowAddAnalysis = this.toggleShowAddAnalysis.bind(this);
+
+        // Set initial state
         this.state = {
```
should this be scrapped too?

swashcap replied:
Doing `this.myMethod = this.myMethod.bind(this)` makes using `this.myMethod` in the `render` method easier. One won’t need to add the `bind` everywhere.

cdaringe replied:
i totally agree.  i was trying to highlight the continued use of `this.state`.

swashcap replied:
Refined form validation in ac102e1a1ffa21f5e0eebf428c7d0a29d5092a3a. This makes the component simpler by expecting its `onSubmit` property.

I looked to the Redux examples to see how they manage user input. The [Todo’s text input](https://github.com/rackt/redux/blob/master/examples/todomvc/components/TodoTextInput.js) delegates internal state to the component. This approach makes sense: an incomplete form isn’t necessarily meaningful application state, although I can imagine wanting to cache user input on an exceptionally long form.

cdaringe commented on app/js/components/consortium-single.js:
```diff
@@ -1,200 +1,191 @@
-            );
-            this.setState(this.state);
+    toggleShowAddAnalysis() {
+        this.setState({
```
😨

cdaringe commented on app/js/components/dashboard-consortia.js:
```diff
@@ -2,16 +2,17 @@
 import React from 'react';
 import ConsortiumCard from './consortium-card';
-import consortia from '../services/consortia';
+import Consortium from '../services/consortium';
```
we've always left Capital-cased (but not FULL_CAPS) for constructor functions/classes.  singletons we've kept camelCased.  this would also impact the casing in services/consortium

swashcap replied:
Good call. I’ll go through and check all references.

cdaringe commented on app/js/middleware/consortium.js:
```diff
@@ -0,0 +1,47 @@
+        default:
+            return nextAction();
+    }
+}
```
1. should we really be piping each action through an additional middleware layer, even if the action is not associated? I did so for auth, but I think Auth **does** implicitly affect every action.
1. what's happening here is:
    1. you've listened to all actions,
    1. detected certain actions that require modifications to the application, but not the SSS directly,
    1. performed those modifications (e.g. async local data persistence), and then
    1.  permitted the simple SSS defined action to proceed

what i like about the approach is that you separate SSS actions from actions that don't update the store.  what i'm not keen on is the need to do it the middleware.  it seems to me that this could be done in the actions.  of course I'm not going to hold this PR up because of it (as you and I are still wrestling on a few topics), but do you feel that there is a rule-of-thumb of when intermediate actions would be middlewared vs. actioned?  My strategy was to do the intermediate action in the controller, then dispatch a fail/success action after the fact.

swashcap replied:
We effectively have *2 stores*: Redux’s store _and_ PouchDB. This is against Redux’s primary philosophy, that of a single source of truth. Furthermore, keeping some state in models adds a _third_ store. This is confusing.

CouchDB is essentially our API. Making calls to an API belongs in action creators. This is fine, but it leads to heavier action creators with a lot of repeated code. I believe action creators should be kept relatively simple (only submitting payloads) to make reasoning about actions simple. The complexity can be moved middleware.

In the case of Consortium, it made sense to modularize to separate concerns, make code more readable and promote reusability. I added a service to make the functionality reusable throughout the application; middleware calls the service.

Here’s what it looks like:

```
+----------------+             +------------+             +---------+
| Action Creator | - action -> | Middleware | - action -> | Reducer |
+----------------+             +------------+             +---------+
                                    |   ^
                                    v   |
                                 +---------+
                                 |   API   |
                                 | Service |
                                 +---------+
                                    |   ^
                                    v   |
                                   +-----+
                                   | API |
                                   +-----+
```

…where “API Service” is the consortium service in this instance.

A lot of the code for consortium operations can be shared. These operations are abstracted in `SingleResponseFactory` and `CollectionResponseFactory` (which shouldn’t be capitalized). 

The reality is that we have at least two sources of truth. The goal with this pattern is to make reasoning about them consistent and understandable.

cdaringe replied:
> We effectively have 2 stores: Redux’s store and PouchDB.

and _nodeapi_, if you are taking a holistic view

>  This is against Redux’s primary philosophy

I disagree.  If that was true, then _any_ external data persistence would be against the philosophy.  The examples in the redux docs calling the _reddit_ API would violate the single source of truth if this were the case.  The application _state_ needs to be represented in a single store, but _not the application data_.  The application data !== application state.

>  some state in models adds a third store

I believe that to make this claim is to not understand their usage.  The single store still holds the truth.  The models _as used_ are an intermediary to reject invalid data, reduce cruft from actions consuming objects, provide I/O formatting, and enable computed values to be put into the store and persistence layer.  They can be lost at app shutdown, then still be regenerated using a previous snapshot of the past app state.  They can be destroyed even after each event handler!  They're really just a tiny intermediary.  Perhaps my holding onto one persistent one in mem in the `project` controller was offputting.  Additionally, my dispatching on events was too magic. **I am OK to discard those practices**.

---

I dont know how you make these epic text things, but that's exactly what they are, epic.

Anyway, I had to go back and read some.  [synchronous `actionCreators` are pure-functions](http://rackt.github.io/redux/docs/basics/Actions.html), but [async actionCreators don't have to be, so long as they're thunks](http://rackt.github.io/redux/docs/advanced/AsyncActions.html).  After giving it some deeper thought, I am OK with it as a practice.  I personally would prefer that these async actions still be branded and live in the `actions` heirarchy, vs. a middleware heirarchy.  Despite that middleware handles the returned thunk, they _are_ actions.  How do you feel about this?  I confirmed that this is how the examples demonstrate their usage.

cdaringe commented on app/js/reducers/consortium.js:
```diff
@@ -0,0 +1,120 @@
+            return Object.assign({}, state, { isLoading: true });
+        case CONSORTIUM_FETCH_ERROR:
+            return Object.assign({}, state, {
+                error: action.error.message,
```
to separate ui state from model state, can we still stick with `ui_error`?

swashcap replied:
Is that an app-wide convention? Where can I see this in action?

cdaringe replied:
i was hoping we could distinguish it as so, until we came up with an alternative.  i use `ui_selectedAnalysis` and `ui_fieldNameState`. in this regard i don't look at `project` and think that `selectedAnalysis` is actually part of an analysis schema

cdaringe commented on app/js/reducers/consortium.js:
```diff
@@ -0,0 +1,120 @@
+        case CONSORTIUM_EDIT_ANALYSIS:
+            /** @todo  Confirm analyses' structure */
+            analyses = state.analyses.map(analysis => {
+                if (analyses.id === action.id) {
```
`analysis.id`?

swashcap replied:
Fixed in 8efc0c08ee18cc42f0bf4b50c4fd455cdc324144.

cdaringe commented on app/js/services/consortium.js:
```diff
@@ -0,0 +1,368 @@
+ *
+ * @return {Promise}
+ */
+function SingleResponseFactory({ consortiumId, key, value }) {
```
this `Factory` updates a consortium based on a kv pair input.  why is that a `Factory`?  Also, what's a `Response`?

cdaringe commented on app/js/services/consortium.js:
```diff
@@ -0,0 +1,368 @@
+ *
+ * @return {Promise}
+ */
+function CollectionResponseFactory({
```
i dont understand what this is or why its needed from brief overview

cdaringe replied:
ok, i see whats going on now.  it's like ur creating service-within-a-service.  you provide a recipe describing your update, and it performs your update.  I didn't find the naming or docblocks descriptive/helpful, respectfully. off hand it seems more convoluted than 'just doing the work' in the respective methods.  however, if it's working, great, and an interesting approach to the problem.  a micro-flux, with the actions and the new state in the same service.  really I think many of the methods are great candidates for putting into the model definitions.  e.g. `myConstortium.addUser(user)`.  it feels more intuitive to me to ask my model to add a user to itself.

cdaringe replied:
maybe this is where our worlds collide.  the services consume the models and dispatch the actions.  this way, it's outside of the controller, like you'd prefer, and out of the middleware, like I'd prefer

cdaringe commented on app/js/services/consortium.js:
```diff
@@ -0,0 +1,368 @@
+     *
+     * @return {Promise}
+     */
+    all() {
```
consortium probably shouldnt proxy consortia.all()

cdaringe replied:
otherwise we're muddying a consortium service and a consortia service

swashcap replied:
Dropped in 9f0762b7b2d4884288e6acdfad8629e06a4cfebd.

cdaringe commented:
cool. looks like a LOT of good thought and effort went into this.  i will let you decide how to best proceed, update, close out, etc.  I would, however, appreciate some feedback on the dialogs that solicited it

swashcap commented:
Okay, I’d say this is ready to merge in. I think all the minor concerns have been addressed (although the commit hashes may be bad because I rebased). The last thing was the form’s `state`, which I cleaned up (see [comment](https://github.com/MRN-Code/coinstac/pull/8#discussion_r41206932)).

We do have some [architectural differences](https://github.com/MRN-Code/coinstac/pull/8#discussion_r40727635) between this feature and the rest of the application. Should we address those before merging in? Thoughts?

cdaringe commented:
1. i think think the ResponseFactory adds more complication that it does assistance, but I'm willing to try it out.  at the least it warrants some more detailed DocBlocks.  it's a service defined within a service, that likely needs to be updated with new actions.
1. at the end our of little discussion on arch, i said "i'm all good."  i just wanted to keep these actions grouped with other actions.  because they are async and exec'd by middleware i thought it would be preferable to keep them in the actions bucket vs. middleware bucket.

please feel free to still merge at your own discretion

swashcap commented:
Added documentation to consortium service's response factories in 51de015282bd70f6afc0c012276ee9632936ffc2.
